apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -845,62 +846,73 @@ do_ensure_all_iterators_closed(_DSSessionID) ->
+                        Stream;
+                    false ->
+                        %% Stream does not exist in mnesia yet, store it.
+                        session_store_stream(SessionId, SubscriptionStream)
+                end
             end,
-            ExistingStreams,
-            Subscriptions
-        )
-    end),
-    ok.
+            SubscriptionStreams
+        ),
+        OrphanedStreams = ExistingStreams -- CurrentStreams,
+        lists:foreach(fun session_delete_stream/1, OrphanedStreams)
+    end).
 
--spec renew_topic_streams(id(), topic_filter_words(), emqx_ds:time(), _Acc :: [ds_stream()]) -> ok.
-renew_topic_streams(DSSessionId, TopicFilter, StartTime, ExistingStreams) ->
-    TopicStreams = emqx_ds:get_streams(?PERSISTENT_MESSAGE_DB, TopicFilter, StartTime),
-    lists:foldl(
-        fun({Rank, Stream}, Streams) ->
-            case lists:keymember(Stream, #ds_stream.stream, Streams) of
-                true ->
-                    Streams;
-                false ->
-                    StreamRef = length(Streams) + 1,
-                    DSStream = session_store_stream(
-                        DSSessionId,
-                        StreamRef,
-                        Stream,
-                        Rank,
-                        TopicFilter,
-                        StartTime
-                    ),
-                    [DSStream | Streams]
-            end
-        end,
-        ExistingStreams,
-        TopicStreams
+-spec get_subscription_streams(session()) ->
+    [{topic_filter(), emqx_ds:time(), {emqx_ds:stream_rank(), emqx_ds:stream()}}].
+get_subscription_streams(#{subscriptions := Subscriptions}) ->
+    subs_fold(fun get_subscription_streams/3, [], Subscriptions).
+
+get_subscription_streams(TopicFilter, #ds_sub{start_time = StartTime}, AccIn) ->
+    TopicFilterWords = emqx_topic:words(TopicFilter),
+    Streams = emqx_ds:get_streams(?PERSISTENT_MESSAGE_DB, TopicFilterWords, StartTime),
+    [{TopicFilter, StartTime, Stream} || Stream <- Streams] ++ AccIn.
+
+find_existing_stream(SubscriptionStream, Existing) ->
+    lists:search(
+        fun(DSStream) -> streams_match(SubscriptionStream, DSStream) end,
+        Existing
     ).
 
-session_store_stream(DSSessionId, StreamRef, Stream, Rank, TopicFilter, StartTime) ->
+streams_match(
+    {TopicFilter, StartTime, {_Rank, Stream}},
+    #ds_stream{topic_filter = TopicFilter, start_time = StartTime, stream = Stream}
+) ->
+    true;
+streams_match(_, _) ->
+    false.
+
+session_store_stream(DSSessionId, {TopicFilter, StartTime, {Rank, Stream}}) ->
     {ok, ItBegin} = emqx_ds:make_iterator(
         ?PERSISTENT_MESSAGE_DB,
         Stream,
-        TopicFilter,
+        emqx_topic:words(TopicFilter),
         StartTime
     ),
     DSStream = #ds_stream{
         session = DSSessionId,
-        ref = StreamRef,
+        ref = emqx_persistent_sequence:next({?SESSION_STREAM_SEQ, DSSessionId}),
```

Comment:
Curious: what are pros and cons against just keeping the counter in session?

Reply:
Pros: slightly more isolated and reusable, a bit less complexity of maintaining a counter (updates don't need to be propagated anywhere), no need to do read/write cycle of the whole session. Cons: one more table brings maintenance burden. To be honest, that was just a quickest way to solve this, given that the data model is in flux we might even drop it quite soon.